desc/compute/utils.py

```diff
@@ -1325,6 +1326,190 @@ def body(i, mins):
+    For a particle with fixed λ, bounce points are defined to be the location
+    on the field line such that the particle's velocity parallel to the
+    magnetic field is zero, i.e. λ |B| = 1.
+
+    Parameters
+    ----------
+    eq : Equilibrium
+        Equilibrium on which the bounce integral is defined.
+    lambdas : ndarray
+        λ values to evaluate the bounce integral at.
+    rho : ndarray
+        Unique flux surface label coordinates.
+    alpha : ndarray
+        Unique field line label coordinates over a constant rho surface.
+    resolution : int
+        Number of quadrature points used to compute the bounce integral.
+
+    Returns
+    -------
+    bi : callable
+        This callable method computes the bounce integral F_ℓ(λ) for every
+        specified field line ℓ (constant rho and alpha), for every λ value in
+        ``lambdas``.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        bi = bounce_integral(eq, lambdas)
+        F = bi(name)
+
+    """
+    if rho is None:
+        rho = jnp.linspace(0, 1, 10)
+    if alpha is None:
+        alpha = jnp.linspace(0, 2 * jnp.pi, 20)
+
+    # Use Gauss-Chebyshev quadrature as the integrand blows up at integration boundary.
+    x, w = chebgauss(deg=resolution)
+    # TODO: Write code to compute bounce points given lambda.
+    #  Vectorize it for multiple lambdas. Then vectorize coordinate mapping logic
+    #  for multiple lambdas. For now, let's pretend bounce points do not depend on
+    #  lambda so that bounce_point() returns either one or two numbers for
+    #  endpoints of all the integrals.
+    bp = bounce_point(eq, lambdas)
+    if bp.size == 1:
+        zeta = -2 * bp * jnp.arcsin(x) / jnp.pi
+    else:
+        zeta = (2 * jnp.arcsin(x) / jnp.pi - 1) / 2 * (bp[1] - bp[0]) + bp[1]
+
+    r, a, z = jnp.meshgrid(rho, alpha, zeta, copy=False, indexing="ij")
+    r, a, z = r.ravel(), a.ravel(), z.ravel()
+    # Now we map these Clebsch-Type field-line coordinates to DESC coordinates.
+    # Note that the rotational transform can be computed apriori because it is a single
+    # variable function of rho, and the coordinate mapping does not change rho. Once
+    # this is known, it is simple to compute theta_PEST from alpha. Then we transform
+    # from straight field-line coordinates to DESC coordinates with the method
+    # compute_theta_coords. This is preferred over transforming from Clebsch-Type
+    # coordinates to DESC coordinates directly with the more general method
+    # map_coordinates. That method requires an initial guess to be compatible with JIT,
+    # and generating a reasonable initial guess requires computing the rotational
+    # transform to approximate theta_PEST and the poloidal stream function anyway.
+    # TODO: In general, Linear Grid construction is not jit compatible.
+    #  This issue can be worked around with a specific routine for this.
+    lg = LinearGrid(rho=rho, M=eq.M_grid, N=eq.N_grid, NFP=eq.NFP, sym=eq.sym)
```

Comment:
Could you explain what is happening here? What is _meshgrid_expand?

Reply:
`meshgrid_expand` is a new function I made that is similar to the functionality as `grid.expand`.

Here I am changing coordinates from field line coordinates $(\rho, \alpha, \zeta)$ to straight field line coords $\rho, \theta_{PEST}, \zeta$. To do that I compute the rotational transform on `lg` - a grid with linearly spaced nodes in DESC coordinates. I would like to compute $\theta_{PEST} = (\alpha + \iota \zeta) \text{mod} (2 \pi)$. To do that I need to now interpolate the rotational transform I computed in DESC coordinates to straight field line coordinates. So first I make an array of iota as a function of rho: `lg.compress(lg_data["iota"])`. Then I call meshgrid expand, which is a general method to upsample `lg.compress(lg_data["iota"])` into an array that is broadcastable with the straight field line grid. Specifically `data["iota"] = _meshgrid_expand(lg.compress(lg_data["iota"]))` will compute the rotational transform on the straight field line grid `iota(\rho, \alpha, \zeta)` so that `(a + data["iota"] * z) % (2 * jnp.pi)` is equivalent to $\theta_{PEST} = (\alpha + \iota \zeta) \text{mod} (2 \pi)$.